modules/dnn/src/webgpu/src/tensor.cpp

```diff
@@ -0,0 +1,129 @@
+#include "../include/tensor.hpp"
+#include "common.hpp"
+#include "internal.hpp"
+#include <unistd.h>
+namespace cv { namespace dnn { namespace webgpu {
+// #ifdef HAVE_WEBGPU
+Tensor::Tensor(Format fmt) : size_in_byte_(0), format_(fmt)
+{
+    createContext();
+    device_ = wDevice;
+}
+
+Tensor::Tensor(const void* data, size_t size_in_byte, 
+                wgpu::BufferUsage usage, Format fmt)
+{
+    createContext();
+    device_ = wDevice;
+    size_in_byte_ = size_in_byte;
+    usage_ = usage;
+    format_ = fmt;
+    setUniform(data, fmt);
+}
+
+Tensor::Tensor(const void* data, std::vector<int>& shape, 
+                wgpu::BufferUsage usage, Format fmt) 
+{
+    createContext();
+    device_ = wDevice;
+    size_in_byte_ = 0;
+    usage_ = usage;
+    format_ = fmt;
+    reshape(data, shape);
+}
+
+const void* Tensor::map()
+{
+    return buffer_->MapReadAsyncAndWait();
+
+}
+
+void Tensor::unMap()
+{
+    buffer_->unMap();
+}
+
+Shape Tensor::getShape() const{
+    return shape_;
+}
+
+int Tensor::count(const int start_axis, const int end_axis) const
+{
+    return shapeCount(shape_, start_axis, end_axis);
+}
+
+int Tensor::dimSize(const int axis) const
+{
+    CV_Assert(axis >= 0);
+    CV_Assert(axis < shape_.size());
+
+    return shape_[axis];
+}
+
+int Tensor::dimNum() const
+{
+    return shape_.size();
+}
+
+Tensor Tensor::reshape(const void* data, const std::vector<int>& shape, 
+                        bool alloc, Format fmt)
+{
+    if (device_ == nullptr)
+    {
+        CV_Error(Error::StsError, "device is NULL");
+        return *this;
+    }
+    CV_Assert(shape.size() > 0 && shape.size() <= 6);
+
+    if (shape_ != shape) shape_ = shape;
+    if (checkFormat(fmt) && fmt != format_) format_ = fmt;
+
+    size_t new_size = shapeCount(shape_) * elementSize(format_);
+    if (alloc || new_size > size_in_byte_)
+        alloc = true;
+    size_in_byte_ = new_size;
+    if (alloc || !buffer_)
+    {
+        // TODO: specific data type 
+        buffer_.reset(new Buffer(device_, data, size_in_byte_, usage_));
+    }
+    else if (data)
+    {
+        buffer_->setBufferData(data, size_in_byte_);
+    }
```

Comment:
duplicated code to setUniform, replace it.

Reply:
Done, thanks for your suggestions, PTAL. @wzw-intel